synapse/media/media_storage.py

```diff
@@ -370,3 +386,240 @@ async def write_chunks_to(self, callback: Callable[[bytes], object]) -> None:
+            self.wrapped_consumer.write(bytes(content_type) + CRLF)
+
+            json_field = json.dumps(self.json_field)
+            json_bytes = json_field.encode("utf-8")
+            self.wrapped_consumer.write(CRLF + json_bytes)
+            self.wrapped_consumer.write(CRLF + b"--" + self.boundary + CRLF)
+
+            self.json_field_written = True
+
+        # if we haven't written the content type yet, do so
+        if not self.content_type_written:
+            type = self.file_content_type.encode("utf-8")
+            content_type = Header(b"Content-Type", type)
+            self.wrapped_consumer.write(bytes(content_type) + CRLF + CRLF)
+            self.content_type_written = True
+
+        self.wrapped_consumer.write(data)
+
+    ### IPushProducer APIs ###
+
+    def stopProducing(self) -> None:
+        """
+        Stop producing data.
+
+        This tells a producer that its consumer has died, so it must stop
+        producing data for good.
+        """
+        assert self.producer is not None
+
+        self.paused = True
+        self.producer.stopProducing()
+
+    def pauseProducing(self) -> None:
+        """
+        Pause producing data.
+
+        Tells a producer that it has produced too much data to process for
+        the time being, and to stop until C{resumeProducing()} is called.
+        """
+        assert self.producer is not None
+
+        self.paused = True
+
+        if self.streaming:
+            cast("interfaces.IPushProducer", self.producer).pauseProducing()
+        else:
+            self.paused = True
+
+    def resumeProducing(self) -> None:
+        """
+        Resume producing data.
+
+        This tells a producer to re-add itself to the main loop and produce
+        more data for its consumer.
+        """
+        assert self.producer is not None
+
+        if self.streaming:
+            cast("interfaces.IPushProducer", self.producer).resumeProducing()
+        else:
+            # If the producer is not a streaming producer we need to start
+            # repeatedly calling  `resumeProducing` in a loop.
+            run_in_background(self._resumeProducingRepeatedly)
+
+    def content_length(self) -> Optional[int]:
+        """
+        Calculate the content length of the multipart response
+        """
+        if not self.length:
+            return None
+        else:
+            # calculate length of json field and content-type header
+            json_field = json.dumps(self.json_field)
+            json_bytes = json_field.encode("utf-8")
+            json_length = len(json_bytes)
+
+            type = self.file_content_type.encode("utf-8")
+            content_type = Header(b"Content-Type", type)
+            type_length = len(bytes(content_type))
+
+            # 154 is the length of the elements that aren't variable, ie
+            # CRLFs and boundary strings, etc
+            self.length += json_length + type_length + 154
+
+            return self.length
```

Comment:
nit: you technically don't need an `else` here since you `return` in the other branch.

```suggestion
        # calculate length of json field and content-type header
        json_field = json.dumps(self.json_field)
        json_bytes = json_field.encode("utf-8")
        json_length = len(json_bytes)

        type = self.file_content_type.encode("utf-8")
        content_type = Header(b"Content-Type", type)
        type_length = len(bytes(content_type))

        # 154 is the length of the elements that aren't variable, ie
        # CRLFs and boundary strings, etc
        self.length += json_length + type_length + 154

        return self.length
```